### PR TITLE
enable template service broker by default

### LINKED ISSUE #15638: Enable TSB by default in 3.7
- treat no config as meaning use openshift namespace
- add Disabled *bool field, nil -> false
- set [Conformance][Feature:TemplateServiceBroker] on relevant extended tests

### PR DESCRIPTION
**Pull request by jim-minter**
fixes #15638

### CODE REVIEW

File: test/extended/setup.sh [hunk 1]
```diff
@@ -102,9 +102,6 @@ function os::test::extended::setup () {
 	cp "${SERVER_CONFIG_DIR}/master/master-config.yaml" "${SERVER_CONFIG_DIR}/master/master-config.orig2.yaml"
 	openshift ex config patch "${SERVER_CONFIG_DIR}/master/master-config.orig2.yaml" --patch="{\"auditConfig\": {\"enabled\": true}}"  > "${SERVER_CONFIG_DIR}/master/master-config.yaml"
 
-	cp "${SERVER_CONFIG_DIR}/master/master-config.yaml" "${SERVER_CONFIG_DIR}/master/master-config.orig2.yaml"
-	openshift ex config patch "${SERVER_CONFIG_DIR}/master/master-config.orig2.yaml" --patch="{\"templateServiceBrokerConfig\": {\"templateNamespaces\": [\"openshift\"]}}"  > "${SERVER_CONFIG_DIR}/master/master-config.yaml"
-
```
Comment (bparees) [inline test/extended/setup.sh]: woo :)

**Comment by bparees** [comment]
lgtm (intentionally not tagging since it needs api review)

**Comment by deads2k** [comment]
In 3.7, we built the infrastructure to install a server as a service via a daemonset.  In 3.8, we created a separate server hosting the template service broker (already done) and just chained it in. 

Why not take advantage of the things we built already and take the last step to split this out before it's too late?  @liggitt

**Comment by bparees** [comment]
> In 3.7, we built the infrastructure to install a server as a service via a daemonset. In 3.8, we created a separate server hosting the template service broker (already done) and just chained it in.
> Why not take advantage of the things we built already and take the last step to split this out before it's too late? @liggitt

@deads2k by 3.7 and 3.8 i assume you mean k8s 1.7 and 1.8 since openshift 3.7+3.8 don't exist yet?

questions:
1) does that  daemonset server still have seamless access to the cluster etcd for resource storage?
2) what are the advantages of running it as a daemonset instead of where it is currently running?
3) is api aggregation seamlessly handled?
4) Of all the various pod controller mechanisms, why a daemonset?
5) why would it be "too late" to do it later?

**Comment by deads2k** [comment]
Sorry, 3.6 and 3.7.  we did the install bits for service catalog and we did the separate server but last week.

It does have access to the cluster etcd (see service catalog).  Daemonset to take advantage or servicecatalog infrastructure.  The API aggregator is always on in 3.7 (already done).  

It's too late to do later because nothing ever moves out of the server after it's on by default and my memory of the template service broker is that it's a one off API (not standard).  It's home should be outside the main API server before it's out of alpha or it will never move.

**Comment by bparees** [comment]
> It does have access to the cluster etcd (see service catalog). Daemonset to take advantage or servicecatalog infrastructure. The API aggregator is always on in 3.7 (already done).

the aggregator being on is not the same as seamless.  There is additional effort involved in registering with it, no?  

> It's too late to do later because nothing ever moves out of the server after it's on by default and my memory of the template service broker is that it's a one off API (not standard). It's home should be outside the main API server before it's out of alpha or it will never move.

The TSB apis (the part that would actually move, since the resource apis would presumably not be moved) are called by the SC, after the TSB tells the SC where it is/how to call it.  I don't really see how moving it later is any more difficult than moving it now.  While historically nothing may have ever moved out of the server once it's on, is there an actual technical reason for that?  Or is it simply a matter of "it became tech debt and people stopped caring?"

**Comment by bparees** [comment]
Also how is configuration handled when it is launched as a daemonset?  For the SC, in cluster up at least, we're hand configuring the deployment that runs it.  I guess the ansible install does the same.  So that's another piece of net new work that needs to be done to move this out, unless there is a seamless way that these daemonsets are given a copy of the master-config automatically.

**Comment by liggitt** [comment]
The master config doesn't seem like the right place to configure an unrelated API server. We should be reducing the number of entangled components, not increasing them. 

Even the logic of the template service broker startup indicates it doesn't belong in the same process as the apiserver, since it requires both the apiserver and controllers before it can run. 

Once it is enabled in process, it becomes extremely difficult to migrate out of process seamlessly across releases. 

We're already seeing the benefit of separate components with the service catalog, which we can update at a different cadence, largely independent of origin and kube versions. As the server intended to be a target of the service catalog, it only makes sense that we'd want the same ability to update independent of origin and kube. 

The tech debt argument is also a strong one, given the pain of the existing things we keep in repo, and the inertia that keeps them in their current state

**Comment by deads2k** [comment]
Opened https://github.com/openshift/origin/pull/15655 to make `openshift start template-service-broker` as a starting point.

**Comment by deads2k** [comment]
https://github.com/openshift/origin/pull/15672 allows the TSB to run separately.

**Pull request closed without merging**